### ISSUE TITLE
`.' inside blank

### ISSUE THREAD

**Issue by wonkydonky** [rdfhdt/hdt-cpp#160]
rdf2hdt is failing to parse Turtle blank nodes like this (example)

        ex:Rome
                ex:founded [
                        ex:when ex:753_B.C. ;
                        ex:where ex:Italy .
                ]

it says

        Input format not given. Guessing from file extension...
        Detected RDF input format: ttl
        error: `.' inside blank                     
        Catch exception load: Error parsing input.
        ERROR: Error parsing input.

basically it says `ex:753_B.C.` is an error because it terminates with a `.`, but the URL should be accepted.
For comparison, `ex:753_B.C` works.

**Comment by wouterbeek** [comment]
Could you be using an older Serd version?  The following succeeds (and so does creating an HDT of the same triple):

```
$ serdi -s "_:a.b <x:x> <x:x> ."
$ serdi -v
serdi 0.29.2 ...
```

**Comment by wonkydonky** [comment]
@wouterbeek Yes I have and older version, `0.28.0`. However `serdi -s "_:a.b <x:x> <x:x> ."` seems to work for me too. What is not working is this one `serdi -s "_:a.b. <x:x> <x:x> ."`. Can you please try this last one?

**Comment by wouterbeek** [comment]
The ending dot is not supposed to work (neither are leading dots BTW).  The relevant grammar line is:

```
[141s] BLANK_NODE_LABEL ::= '_:' (PN_CHARS_U | [0-9]) ((PN_CHARS | '.')* PN_CHARS)?
```

**Comment by wonkydonky** [comment]
Alright. The absolute URLs are valid though, so there should probably be a way to express them even when serialized using a blank node or a prefix.

**Comment by wouterbeek** [comment]
Unfortunately, it does not seem possible to escape leading and/or trailing dots in blank node labels in Turtle.  However, this does not restrict expressiveness since blank node labels are arbitrary.  IOW, there is no guarantee that a blank node label with a leading and/or trailing dot will be preserved under loading and saving the data.

**Comment by wonkydonky** [comment]
It might be a problem for non blank nodes however, because this also is invalid `serdi -s "prefix ex: <http://example.org/> ex:a.b. <x:x> <x:x> ."` so it looks like there is no way to write this URLs in Turtle using a prefix.

**Comment by wouterbeek** [comment]
Thanks for sharing the example, now I understand what you mean :-)

I think you're on to something here; the following should result in a valid parse I think:

```
$ serdi -s "prefix : <x:x> :\. : : ."
error: (string):1:20: bad subject
```

I'm going to open this issue in the Serd project.

**Comment by wonkydonky** [comment]
Indeed, looking at [here](https://www.w3.org/TR/turtle/#sec-escapes) what I understand is that "reserved characters" should be escaped, therefore `ex:a\.b\.` should be accepted.

Thanks for opening the issue at serd.

**Comment by wouterbeek** [comment]
This was fixed in Serd some time ago, see https://github.com/drobilla/serd/issues/13
